### ISSUE TITLE
GGUF endianness cannot be determined from GGUF itself

### ISSUE THREAD

**Issue by philpax** [ggerganov/llama.cpp#3957]
As of the time of writing, the big-endian support that was added in https://github.com/ggerganov/llama.cpp/pull/3552 doesn't encode the endianness within the file itself: 

https://github.com/ggerganov/llama.cpp/blob/3d48f42efcd05381221654376e9f6f69d76af739/gguf-py/gguf/gguf.py#L689-L698

This means that there is no way to distinguish a big-endian GGUF file from a little-endian file, which may cause some degree of consternation in the future if these files get shared around 😅 

The cleanest solution would be to add the endianness to the header - ideally, it would be in the metadata, but the reading of the metadata is dependent on the endianness - but that would be a breaking change.

Given that, my suggestion would be to use `FUGG` as the header for big-endian files so that a little-endian executor won't attempt to read it at all unless it knows how to deal with it. The same can go the other way, as well (a big-endian executor won't attempt to read a little-endian executor).

**Comment by KerfuffleV2** [comment]
> Given that, my suggestion would be to use `FUGG`

Seems appropriate since that's probably also what you'd say if you tried to use a model of the wrong endianness.

**Comment by abrander** [comment]
> This means that there is no way to distinguish a big-endian GGUF file from a little-endian file, which may cause some degree of consternation in the future if these files get shared around 😅

Well. The future is now :smile: 

> The cleanest solution would be to add the endianness to the header - ideally, it would be in the metadata, but the reading of the metadata is dependent on the endianness - but that would be a breaking change.

I suggest storing the endianness as a single byte that can be read without ambiguity on both little- and big-endian hosts.

Reading a 16-, 32- or 64-bit number is often confusing when you don't know the endianness of the file itself. It quickly leads to very confusing code where you must first guess the host byte order, whether you're reading the file using native byte order, little-endian, or big-endian byte order, and somehow combine those using pure magic and `#ifdef` guards.

We could use a single int8 (The ASCII value `L` or `B` for easy detection by humans in a hex viewer) to distinguish the file byte order.

It will be a breaking change, but failure because of hidden endian issues is also a breaking change.

> Given that, my suggestion would be to use `FUGG` as the header for big-endian files so that a little-endian executor won't attempt to read it at all unless it knows how to deal with it. The same can go the other way, as well (a big-endian executor won't attempt to read a little-endian executor).

The only argument I see for making the file magic endian dependable would be to make incompatible readers bail fast - but hey, if they ignore the version field, all bets are off anyway.

**Comment by KerfuffleV2** [comment]
Hmm, does anything actually need to change other than just storing the version field in the respective endianness? Until there's GGUF version 20,000 or something, every big-endian version is going to be invalid for little endian and every little endian version is going to be invalid for big endian. (Except 0 which is the same for both, but we're already past that.)

**Comment by philpax** [comment]
Hm - yes, you're both right, I hadn't noticed the version was written with the correct endianness. This is then not as much of an issue as it seemed - big-endian files that end up out there are unlikely to be loadable by existing executors if they're correctly version-checking.

That being said, I'd still prefer to make it explicit instead of having it be an implicit property of the version number. I can't imagine it'll ever reach a version number in which there's a realistic possibility of confusion, but it's easier for readers if they can determine it directly without having to check if they support the endian-swapped version number.

> _edit_: If you don't like the idea of just relying on the version being invalid if the file was created for the wrong endian, an alternative (also backward compatible except for existing BE files) is to say the version field is still 4 bytes but the value is a `uint16` now. LE can use the first two bytes - `uint32` value `4` is `\x04\x00\x00\x00` and `uint16` value `4` is `\x04\x00` so no change is needed. BE can use the second two bytes. Assuming existing BE files wrote the version in BE format, then that also will already be correct since BE `4` is `\x00\x00\x00\x04` and BE `uint16` `[0,4]` would be the same.

Interesting and clever idea. This could work and wouldn't require a strictly-breaking change to the format.

The immediate concerns I have are that:
- no matter what you do, you'll be wasting two bytes. This really doesn't matter given the models are gigabytes in size, but figured I'd mention it still.
- the writers and readers need to be aware that they need to check the last two bytes instead of the first two if running on BE. 

  This isn't a deal-breaker, but it seems pretty easy to forget to do this. To be fair, most people aren't going to be developing against big-endian anyway - perhaps it's fine if it just breaks there.

Ultimately I don't think either of these really matter, it's just not quite as elegant as a more dedicated detection mechanism. I'd be fine with this if we wanted to maintain strict backwards compatibility.

---

With that being said, a cursory search indicates that nobody's uploaded big-endian models to HF yet, so I'm pretty comfortable with breaking them specifically as long as LE's fine. I can think of two potential solutions here:

- Change the magic number based on endianness (original proposal). This makes it *very* obvious, but may confuse people looking at the format in the wild if they've never seen the big-endian variant before.
- Encode the endianness in the version number in a way that's freely compatible with LE and easy to implement for BE. The easy solution is to make it a right-aligned `uint16` and use the first two bytes as a boolean (first byte is a big-endian-active bool, second byte is unused). 

  If we want to be less wasteful and allow for going past 65,536, we can encode it in the MSB of the first byte, but this requires an annoying amount of masking for what seems like a very unlikely outcome.

I'm leaning towards the right-aligned uint16 solution because it maintains LE compatibility, is still obviously a GGUF file, and should work for us until sometime in the next century. 

I could be convinced to use one of the three other solutions (magic number solution, using only one bit, or BE-backwards-compatible moving uint16), but this seems like the best set of tradeoffs.

**Comment by KerfuffleV2** [comment]
If anyone wants to test with non-native endian files, I added a conversion script in #3981 - you can just do something like:

    python convert-gguf-endian.py blah.gguf big

**Warning**: This does not produce a new file, it modifies the input.

It can only convert `F32`, `F16` and `Q8_0` tensors.

**Comment by KerfuffleV2** [comment]
@philpax 

> If we want to be less wasteful and allow for going past 65,536

I don't mind leaving the problem of dealing with GGUF version 65536 to the people in the year 4,000.

My solution was actually a little overcomplicated though. You can just read the field as normal then bitwise AND with `65535` and you'll get `0` if it's the wrong endian.

**Comment by abrander** [comment]
For anyone interested, I have implemented v3 byte order detection - without needing to know the host byte order - using the last byte of the version field as a big-endian marker.

It works as intended. It's a bit ugly, but it works.

**Comment by KerfuffleV2** [comment]
> It works as intended. It's a bit ugly, but it works.

Until version 256. :) That still may never happen in our lifetimes, but still. I think `& 65535` is a bit more reliable. Then if it's 0, you can byte swap and test if it's a valid version. (The approach I took.)

**Comment by philpax** [comment]
Yeah I think I'd be fine with `& 65536`. That seems pretty straightforward and I will be impressed if we manage to get anywhere near that in our lifetimes.

**Comment by KerfuffleV2** [comment]
> Yeah I think I'd be fine with `& 65536`. 

I don't think you will be. :) It's gotta be `65535` here unless you want `0` and `65536` as the only two possible values.

**Comment by philpax** [comment]
Yes, of course - my bad 😅

**Comment by julien-c** [comment]
Thanks @ggerganov for pointing me to this nice convo, it should enable us to support visualization of big-endian gguf files in HF: 
see https://github.com/huggingface/huggingface.js/pull/540 (and the small fix https://github.com/huggingface/huggingface.js/pull/545) for context! 🔥

**Comment by philpax** [comment]
Just noticed this was marked as stale - as far as I know, this is still an issue, the only indicator that a GGUF is big-endian is if its version is in big-endian, which is not a strong signal. Given that big-endian models are the minority by far, I'd continue to suggest reversing the magic number to make it explicit, and forcing a re-encode of the BE models.

**Comment by ggerganov** [comment]
> and forcing a re-encode of the BE models.

Re-encode "GGUF" -> "FUGG", is this the proposal?

**Comment by philpax** [comment]
Yes, pretty much. I think that should be sufficient?

**Comment by julien-c** [comment]
> the only indicator that a GGUF is big-endian is if its version is in big-endian, which is not a strong signal

Seems to work well enough in existing implementations, though (just my 2 cents)

**Comment by philpax** [comment]
It works as a weak signal ("this file can't be loaded"), but you have to use a heuristic to determine if it's a big-endian file ("could I support this if the version was endian-swapped?"). 

A reader could determine if a file is big-endian and provide a more informative error in that case; another reader could also offer to swap the bytes around for the user. This could also be useful for online viewers, which could definitively tell you what type of model it is before you download it.

In general, I think it's a good idea to make things like this as upfront as possible in the reading process so that a reader knows what it's working with without having to intuit it with a heuristic.

**Comment by Scottcjn** [comment]
I would like to take this on. I run an IBM POWER8 S824 (big-endian ppc64) with llama.cpp daily and have been working on POWER8 optimizations (VSX vec_perm, NUMA-aware weight loading, IBM MASS integration).

To confirm my understanding of the proposal:

1. **Writer**: When producing a big-endian GGUF file, write `FUGG` (bytes `0x46 0x55 0x47 0x47`) as the magic instead of `GGUF` (`0x47 0x47 0x55 0x46`)
2. **Reader**: Accept both `GGUF` (little-endian) and `FUGG` (big-endian) as valid magic, using the magic itself to determine byte order instead of the version-field heuristic
3. **Endian converter**: When converting LE→BE, also swap the magic from `GGUF`→`FUGG` (and vice versa)
4. **Test update**: `tests/test-gguf.cpp` currently uses `FUGG` as "bad magic" — this would need a different invalid magic

Files I plan to touch:
- `ggml/include/gguf.h` — add `GGUF_MAGIC_BE`
- `ggml/src/gguf.cpp` — reader accepts both magics, writer selects based on endianness
- `gguf-py/gguf/constants.py` — add `GGUF_MAGIC_BE`
- `gguf-py/gguf/gguf_reader.py` — detect endianness from magic directly
- `gguf-py/gguf/gguf_writer.py` — write correct magic per target endianness
- `gguf-py/gguf/scripts/gguf_convert_endian.py` — swap magic during conversion
- `tests/test-gguf.cpp` — update bad magic test case

I can test this end-to-end on actual big-endian hardware (POWER8 S824, 512GB RAM, Ubuntu 20.04). Will keep the version-field heuristic as a fallback for reading legacy BE files that still have `GGUF` magic.

Will have a PR up soon.

**Comment by Scottcjn** [comment]
I'd like to take this. We run the only active big-endian (POWER8 ppc64) llama.cpp deployment and can test both reading and writing BE GGUF files on real hardware.

**Proposed approach** (based on the discussion above):

1. Use `FUGG` (0x46554747) as the magic header for big-endian GGUF files (reversed `GGUF` = 0x47475546)
2. When reading: check magic — if `GGUF` → little-endian, if `FUGG` → big-endian, else reject
3. When writing on BE host: write `FUGG` magic + BE-native field ordering
4. The existing `gguf-convert-endian.py` script would swap magic as part of conversion
5. Backward compat: existing BE files without `FUGG` magic are still loadable via the version-check heuristic @KerfuffleV2 described

This is a clean, minimal change to `ggml.c` (header read/write) + `gguf-py` (Python writer). Will submit a PR with tests on both LE (x86) and BE (POWER8) outputs.

Any concerns before I start?

**Comment by Clypeum26** [comment]
Hi, completely new to open source contributions - used git only on my laptop with my personal repos, so never have had to open up a pull request before.
**Yes, I have read the CONTRIBUTING file and the README** pointed me to these issues since they are supposedly easy to do.
This one seems to be easy enough even though C++ is not really a language I know well (however, *C* is a love child for me and most languages are the same anyways and nothing a bit of googling can't fix).


Main thing I want to know is if this is an actual issue since [taronaeo](https://github.com/taronaeo) said the following:


> Also, Big-Endian (BE) GGUFs have been supported for awhile now with llama-cli throwing warning messages if heuristics 
> determine that the byte-order is incorrect.
> 
> I'm not sure if this is still an "issue" to be fixed.


It was on [this previously closed "for being a pure ai repo" thingy](https://github.com/ggml-org/llama.cpp/pull/20019).

Basically - does it need to be fixed or is the version checking thing fine as it is now?


Also another question/s:
Could you elaborate on how this is a "breaking change"?

My guess would be that now it would break on older BE (*Big Endian*) files with GUFF as their magic?
This is kind of the only thing that comes to mind.

If that is the case I'm guessing it would be preferable to have both the GUFF/FUGG magic label and also the heuristic version check as a failsafe?

In addition, the ai above said stuff about the test files using FUGG as the wrong magic example - does that also need to be changed (I know that tests are a big no-no zone for people who don't know what the tests actually test)?

(Edit - I am running a Mac which from my slight googling is a LE enjoyer, so I wouldn't be able to test if this would 100% work out in the wild. Is this ok? I would still try to have my own tests to see if it detects it correctly, but it wouldn't really be a trial by fire.)

Thanks to whoever is reading this (if there even is anyone who looks at this).
P.S. Final count of writing "FFUG" instead of "FUGG"/"GUFF" = 6 (There aren't even that many spaces where I wrote it).
P.P.S. I am quite sure that the markdown will look weird since I do not know how github will render this exactly.
(Edit - found out that ` is not the same as >)
